### PR TITLE
[Validator] Ensure nested validations for Collections happen in a separate context

### PR DESCRIPTION
**Pull request by evanpurkhiser**
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #14639 |
| License | MIT |
| Doc PR | n/a |

### CODE REVIEW

**Comment by EvanPurkhiser** [comment]
Ping @fabpot

**Comment by fabpot** [comment]
ping @symfony/deciders

**Comment by stof** [comment]
I would like a validation of @webmozart to be sure it is the right fix. I'm not sure the ExecutionContext is meant to be clonable (and different implementations might not share the same ViolationList when cloning btw, which would break this fix)

**Comment by EvanPurkhiser** [comment]
Hmm, I think this *may* have some relation to https://github.com/symfony/symfony/pull/11412. Looking at the changes in that fix, indeed maybe cloning the context might not be the right solution.

Hopefully @webmozart can comment :-)

**Comment by fabpot** [comment]
@webmozart Can you tell us your point of view on this PR?

**Comment by xabbuh** [comment]
ping @webmozart

**Comment by reket** [comment]
ping @webmozart

**Comment by EvanPurkhiser** [comment]
Ping @webmozart on this again :-)

**Comment by fabpot** [comment]
@webmozart The change made in this PR is "simple" enough but we really need your help to validate that this is indeed the right way.

File: src/Symfony/Component/Validator/Constraints/CollectionValidator.php [hunk 1]
```diff
@@ -60,7 +60,7 @@ public function validate($value, Constraint $constraint)
                 if (count($fieldConstraint->constraints) > 0) {
                     if ($context instanceof ExecutionContextInterface) {
                         $context->getValidator()
-                            ->inContext($context)
+                            ->inContext(clone $context)
```
Comment (webmozart) [inline src/Symfony/Component/Validator/Constraints/CollectionValidator.php]: This solution doesn't work. Now all constraints thrown by the validators in the `Collection` constraint will be added to the cloned context and not appear in the current context (which is the point of calling `inContext()`).

**Comment by webmozart** [comment]
@fabpot Thanks for the ping. Unfortunately this solution would bringe more problems than it would solve. The proper solution is to reset the constraint in the ExecutionContext to the previous one after the nested validation call. We already [do the same for all other context properties](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php#L114).

We need to add a new method `getConstraint()` to ExecutionContext in order to do this. When we switch to 4.0, the method should be added to ExecutionContextInterface as well.

**Comment by webmozart** [comment]
Status: Needs Work

**Comment by fabpot** [comment]
Closing as @webmozart mentioned that this is not something we can merge anyway. As there is a corresponding issue, this will still be linked there in case someone want to work on @webmozart suggestion.

**Comment by xabbuh** [comment]
see #20664 for an alternative approach

**Pull request closed without merging**